### ISSUE TITLE
[Feature]: Integrating user information generated by LiteLLM with Langfuse

### ISSUE THREAD

**Issue by vrvrv** [BerriAI/litellm#7238]
### The Feature
I would like to propose a feature where, when a request is made to LiteLLM using a user API token issued by LiteLLM, the user ID associated with that token is automatically logged. Additionally, it would be beneficial if this information could be automatically logged in Langfuse to identify which user created the trace.
This integration will allow the automatic inclusion of user_id in the logging process when requests are sent to Langfuse using the issued tokens. This ensures that Langfuse can accurately track and differentiate request volumes on a per-user basis.

### Motivation, pitch

I need accurately log and analyze user-specific spend and token usage in Langfuse. Currently, without the `trace_user_id` being included in metadata manually, it is challenging to distinguish between requests made by different users, which limits the ability to perform user-specific analytics and reporting (when user misses including metadata)

By enforcing the integration of user information from LiteLLM into Langfuse, we can ensure that each request is associated with the correct userID. This feature is crucial for projects that require detailed user activity tracking and analysis, such as those involving personalized services or usage-based billing.

This proposal addresses the problem of insufficient user-specific logging and enhances the overall functionality and utility of both LiteLLM and Langfuse systems.

### Are you a ML Ops Team?

No

**Comment by ishaan-jaff** [comment]
- what information do you want to see ? 
- should litellm auto add this based on the virtual key used ? 
- Ideally can you write out all fields you want litellm to fill in ?

**Comment by ghost** [comment]
> using a user API token issued by LiteLLM, the user ID associated with that token is automatically logged

Hey @vrvrv this is already done - https://docs.litellm.ai/docs/proxy/logging_spec#standardlogginguserapikeymetadata

see: `user_api_key_user_id`

**Comment by castrapel** [comment]
Currently, Langfuse traces are not associated with users even when including `user_api_key_user_id` in `langfuse_default_tags`. This results in fragmented conversation traces in LangFuse that are not attributable to a user. Yes, the user is logged in LangFuse as a tag, but this is not attributable in the way Langfuse expects. When using LiteLLM's LangFuse integration, I see the following in Traces (Notice no user attribution, even though it is present in the tags):

![Image](https://github.com/user-attachments/assets/5c82c10b-300a-49cf-a1bc-d80f24d89058)

As a result of this, the list of Sessions are empty.

In a custom integration we have, we pass the user_id parameter explicitly via the LangFuse API, and users appear like this:

![Image](https://github.com/user-attachments/assets/18c27e17-6755-4b53-8621-11f75bd08016)

As a result, LangFuse compiles the session of all the user's messages in a given conversation properly, and the Sessions page is not empty:

![Image](https://github.com/user-attachments/assets/bef54be5-a786-4781-807b-e39eff298215)
 
Here's an example of how we attribute the user_id to the user's email address, the key part being the `user_id` field:

```python
            trace_payload = {
                "name": f"chat:{chat_id}",
                "input": body.get("messages", [])[0] if body.get("messages") else None, # Log first message as input
                "user_id": user_email,
                "metadata": metadata,
                "session_id": chat_id,
            }
            if tags_list:
                trace_payload["tags"] = tags_list

            try:
                trace = self.langfuse.trace(**trace_payload)
...
```

If helpful, I can provide an example of what would be provided in each of those fields.

**Comment by ghost** [comment]
hey @castrapel do you see `user` via litellm, if you pass one in the request body?

**Comment by castrapel** [comment]
I think you're referring to making a manual request to litellm with the `user` tag when I call the OpenAI endpoint, and yes, that worked. I'd expect this to be added to the request to LangFuse automatically by LiteLLM, since the api key I am using with openai is associated with the user in litellm, and user in litellm already has an email address assigned to them. By making the user responsible for including their email address, we completely trust the client to provide whichever email they want. Shouldn't LiteLLM Proxy automatically supply the user email address to LangFuse, and prevent the user from providing an arbitrary, potentially forged, email address?

Also, it looks like we also need the session_id payload in order to force LangFuse to create a session. Our code (which is completely separate from LiteLLM) just generates a session ID for the user's conversation. I think the session ID can be provided by the caller, but if not, LiteLLM might want to generate a UUID and pass that as the session_id for LangFuse.

**Comment by ghost** [comment]
Hey @castrapel so there's 2 users on litellm 

- internal user (person that owns the key) 
- end user (end user of an ai service) 

the end user can be passed in via the `user=` param which you just tried. 

Langfuse - i don't think - has that distinction, so how would you want us to handle this?

**Comment by castrapel** [comment]
@krrishdholakia , you are correct. The caller should define the user, given that LiteLLM is treated as middleware much of the time, and the LiteLLM "internal user" is passed as a tag.

In that case, I think the only other issue is being able to pass a Session ID?

**Comment by ghost** [comment]
Hey @castrapel are you unable to pass the session id today? 

I can see we support it - https://github.com/BerriAI/litellm/blob/298a3574f4ceecae02bde28d671b740a1e777e65/litellm/integrations/langfuse/langfuse.py#L503
